File: _includes/code/csharp/ManageObjectsImportTest.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using CsvHelper;
using Weaviate.Client;
using Weaviate.Client.Models;
using Xunit;

namespace WeaviateProject.Tests;

[Collection("Sequential")]
public class ManageObjectsImportTest : IAsyncLifetime
{
    private static readonly WeaviateClient client;
    private const string JsonDataFile = "jeopardy_1k.json";
    private const string CsvDataFile = "jeopardy_1k.csv";

    // Static constructor for one-time setup (like @BeforeAll)
    static ManageObjectsImportTest()
    {
        // START INSTANTIATION-COMMON
        string openaiApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
        if (string.IsNullOrWhiteSpace(openaiApiKey))
        {
            throw new ArgumentException("Please set the OPENAI_API_KEY environment variable.");
        }

        client = Connect.Local().GetAwaiter().GetResult();
        // END INSTANTIATION-COMMON
    }

    // A helper method to generate a deterministic UUID from a seed
    private static Guid GenerateUuid5(string seed)
    {
        var namespaceId = Guid.Empty;
        var namespaceBytes = namespaceId.ToByteArray();
        var nameBytes = Encoding.UTF8.GetBytes(seed);

        var combinedBytes = new byte[namespaceBytes.Length + nameBytes.Length];
        Buffer.BlockCopy(namespaceBytes, 0, combinedBytes, 0, namespaceBytes.Length);
        Buffer.BlockCopy(nameBytes, 0, combinedBytes, namespaceBytes.Length, nameBytes.Length);

        using (var sha1 = SHA1.Create())
        {
            var hash = sha1.ComputeHash(combinedBytes);
            var newGuid = new byte[16];
            Array.Copy(hash, 0, newGuid, 0, 16);

            newGuid[6] = (byte)((newGuid[6] & 0x0F) | (5 << 4));
            newGuid[8] = (byte)((newGuid[8] & 0x3F) | 0x80);

            (newGuid[0], newGuid[3]) = (newGuid[3], newGuid[0]);
            (newGuid[1], newGuid[2]) = (newGuid[2], newGuid[1]);
            (newGuid[4], newGuid[5]) = (newGuid[5], newGuid[4]);
            (newGuid[6], newGuid[7]) = (newGuid[7], newGuid[6]);

            return new Guid(newGuid);
        }
    }

    // Runs once before any tests in the class
    public async Task InitializeAsync()
    {
        using var httpClient = new HttpClient();
        var jsonData = await httpClient.GetStreamAsync(
            "https://raw.githubusercontent.com/weaviate-tutorials/edu-datasets/main/jeopardy_1k.json"
        );
        using var fileStream = new FileStream(JsonDataFile, FileMode.Create, FileAccess.Write);
        await jsonData.CopyToAsync(fileStream);
    }

    // Runs once after all tests in the class
    public async Task DisposeAsync()
    {
        await client.Collections.DeleteAll();
        File.Delete(JsonDataFile);
        if (File.Exists(CsvDataFile))
            File.Delete(CsvDataFile);
    }

    private async Task BeforeEach()
    {
        await client.Collections.DeleteAll();
    }

    [Fact]
    public async Task TestBasicBatchImport()
    {
        await BeforeEach();
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "MyCollection",
                VectorConfig = Configure.Vector("default", v => v.SelfProvided()),
            }
        );

        // START BasicBatchImportExample
        var dataRows = Enumerable
            .Range(0, 5)
            .Select(i => new { title = $"Object {i + 1}" })
            .ToList();

        var collection = client.Collections.Use("MyCollection");

        // highlight-start
        var response = await collection.Data.InsertMany(dataRows);
        // highlight-end

        var failedObjects = response.Where(r => r.Error != null).ToList();
        if (failedObjects.Any())
        {
            Console.WriteLine($"Number of failed imports: {failedObjects.Count}");
            Console.WriteLine($"First failed object: {failedObjects.First().Error}");
        }
        // END BasicBatchImportExample

        var result = await collection.Aggregate.OverAll(totalCount: true);
        Assert.Equal(5, result.TotalCount);
    }

    // START ServerSideBatchImportExample
    // Coming soon
    // END ServerSideBatchImportExample

    [Fact]
    public async Task TestBatchImportWithID()
    {
        await BeforeEach();
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "MyCollection",
                VectorConfig = Configure.Vector("default", v => v.SelfProvided()),
            }
        );

        // START BatchImportWithIDExample
        var dataToInsert = new List<BatchInsertRequest>();
        var vectorData = Enumerable.Repeat(0.1f, 10).ToArray();

        for (int i = 0; i < 5; i++)
        {
            var dataRow = new { title = $"Object {i + 1}" };
            var objUuid = GenerateUuid5(JsonSerializer.Serialize(dataRow));

            var vectors = new Vectors { { "default", vectorData } };

            dataToInsert.Add(
                BatchInsertRequest.Create(data: dataRow, uuid: objUuid, vectors: vectors)
            );
        }

        var collection = client.Collections.Use("MyCollection");

        // highlight-start
        var response = await collection.Data.InsertMany(dataToInsert);
        // highlight-end

        var failedObjects = response.Where(r => r.Error != null).ToList();
        if (failedObjects.Any())
        {
            Console.WriteLine($"Number of failed imports: {failedObjects.Count}");
            Console.WriteLine($"First failed object: {failedObjects.First().Error}");
        }
        // END BatchImportWithIDExample

        Assert.Empty(failedObjects);

        var result = await collection.Aggregate.OverAll(totalCount: true);
        Assert.Equal(5, result.TotalCount);

        var lastUuid = dataToInsert[4].UUID;
        Assert.NotNull(await collection.Query.FetchObjectByID((Guid)lastUuid));
    }

    [Fact]
    public async Task TestBatchImportWithVector()
    {
        await BeforeEach();
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "MyCollection",
                VectorConfig = Configure.Vector("default", v => v.SelfProvided()),
            }
        );

        // START BatchImportWithVectorExample
        var dataToInsert = new List<BatchInsertRequest>();
        var vectorData = Enumerable.Repeat(0.1f, 10).ToArray();

        for (int i = 0; i < 5; i++)
        {
            var dataRow = new { title = $"Object {i + 1}" };
            var objUuid = GenerateUuid5(JsonSerializer.Serialize(dataRow));

            var vectors = new Vectors { { "default", vectorData } };

            dataToInsert.Add(
                BatchInsertRequest.Create(data: dataRow, uuid: objUuid, vectors: vectors)
            );
        }

        var collection = client.Collections.Use("MyCollection");

        var response = await collection.Data.InsertMany(dataToInsert);

        // Handle errors
        if (response.HasErrors)
        {
            Console.WriteLine($"Number of failed imports: {response.Errors.Count()}");
            Console.WriteLine($"First failed object: {response.Errors.First().Message}");
        }
        // END BatchImportWithVectorExample

        var result = await collection.Aggregate.OverAll(totalCount: true);
        Assert.Equal(5, result.TotalCount);
    }

    [Fact]
    public async Task TestBatchImportWithCrossReference()
    {
        await BeforeEach();
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "Publication",
                Properties = [Property.Text("title")],
            }
        );
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "Author",
                Properties = [Property.Text("name")],
                References = [new Reference("writesFor", "Publication")],
            }
        );

        var authors = client.Collections.Use("Author");
        var publications = client.Collections.Use("Publication");

        var fromUuid = await authors.Data.Insert(new { name = "Jane Austen" });
        var targetUuid = await publications.Data.Insert(new { title = "Ye Olde Times" });

        // START BatchImportWithRefExample
        var collection = client.Collections.Use("Author");

        var response = await collection.Data.ReferenceAddMany([
            new DataReference(fromUuid, "writesFor", targetUuid),
        ]);

        if (response.HasErrors)
        {
            Console.WriteLine($"Number of failed imports: {response.Errors.Count}");
            Console.WriteLine($"First failed object: {response.Errors.First()}");
        }
        // END BatchImportWithRefExample

        var result = await collection.Query.FetchObjectByID(
            fromUuid,
            returnReferences: [new QueryReference("writesFor")]
        );
        Assert.NotNull(result);
        Assert.True(result.References.ContainsKey("writesFor"));
    }

    [Fact]
    public async Task TestImportWithNamedVectors()
    {
        await BeforeEach();
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "MyCollection",
                VectorConfig = new[]
                {
                    Configure.Vector("title", v => v.SelfProvided()),
                    Configure.Vector("body", v => v.SelfProvided()),
                },
                Properties = [Property.Text("title"), Property.Text("body")],
            }
        );

        // START BatchImportWithNamedVectors
        var dataToInsert = new List<BatchInsertRequest>();

        for (int i = 0; i < 5; i++)
        {
            var dataRow = new { title = $"Object {i + 1}", body = $"Body {i + 1}" };
            var titleVector = Enumerable.Repeat(0.12f, 1536).ToArray();
            var bodyVector = Enumerable.Repeat(0.34f, 1536).ToArray();

            // highlight-start
            var namedVectors = new Vectors { { "title", titleVector }, { "body", bodyVector } };

            dataToInsert.Add(BatchInsertRequest.Create(dataRow, vectors: namedVectors));
            // highlight-end
        }

        var collection = client.Collections.Use("MyCollection");

        // Insert the data using InsertMany
        // highlight-start
        var response = await collection.Data.InsertMany(dataToInsert);
        // highlight-end

        // Handle errors
        if (response.HasErrors)
        {
            Console.WriteLine($"Number of failed imports: {response.Errors.Count()}");
            Console.WriteLine($"First failed object error: {response.Errors.First().Message}");
        }
        // END BatchImportWithNamedVectors
    }

    [Fact]
    public async Task TestJsonStreaming()
    {
        await BeforeEach();
        // Ensure using correct Collection creation syntax
        await client.Collections.Create(
            new CollectionCreateParams
            {
                Name = "JeopardyQuestion",
                // Optional: Define properties explicitly if needed, but auto-schema usually handles it
                Properties = [Property.Text("question"), Property.Text("answer")],
            }
        );

        // START JSON streaming
        int batchSize = 100;
        var batch = new List<object>(batchSize);
        var collection = client.Collections.Use("JeopardyQuestion");

        Console.WriteLine("JSON streaming, to avoid running out of memory on large files...");
        using var fileStream = File.OpenRead(JsonDataFile);

        // Deserialize as JsonElement to handle types more safely/explicitly than Dictionary<string, object>
        var jsonObjects = JsonSerializer.DeserializeAsyncEnumerable<JsonElement>(fileStream);

        await foreach (var obj in jsonObjects)
        {
            // JsonElement is a struct, checking ValueKind is safer than null check
            if (obj.ValueKind == JsonValueKind.Null || obj.ValueKind == JsonValueKind.Undefined)
                continue;

            var properties = new
            {
                question = obj.GetProperty("Question").ToString(),
                answer = obj.GetProperty("Answer").ToString(),
            };

            batch.Add(properties);

            if (batch.Count == batchSize)
            {
                await collection.Data.InsertMany(batch);
                Console.WriteLine($"Imported {batch.Count} articles...");
                batch.Clear();
            }
        }

        if (batch.Any())
        {
            await collection.Data.InsertMany(batch);
            Console.WriteLine($"Imported remaining {batch.Count} articles...");
        }

        Console.WriteLine("Finished importing articles.");
        // END JSON streaming

        var result = await collection.Aggregate.OverAll(totalCount: true);
        Assert.Equal(1000, result.TotalCount);
    }

    [Fact]
    public async Task TestCsvStreaming()
    {
        await BeforeEach();
        // Create a CSV file from the JSON for the test
        using (var fileStream = File.OpenRead(JsonDataFile))
        using (var writer = new StreamWriter(CsvDataFile))
        using (var csv = new CsvWriter(writer, CultureInfo.InvariantCulture))
        {
            var jsonObjects = JsonSerializer.DeserializeAsyncEnumerable<Dictionary<string, object>>(
                fileStream
            );
            csv.WriteHeader<JeopardyQuestion>();
            await csv.NextRecordAsync();
            await foreach (var obj in jsonObjects)
            {
                if (obj != null)
                {
                    csv.WriteRecord(
                        new JeopardyQuestion
                        {
                            Question = obj["Question"]?.ToString(),
                            Answer = obj["Answer"]?.ToString(),
                        }
                    );
                    await csv.NextRecordAsync();
                }
            }
        }

        await client.Collections.Create(new CollectionCreateParams { Name = "JeopardyQuestion" });

        // START CSV streaming
        int batchSize = 100;
        var batch = new List<object>(batchSize);
        var collection = client.Collections.Use("JeopardyQuestion");

        Console.WriteLine("CSV streaming to not load all records in RAM at once...");
        using (var reader = new StreamReader(CsvDataFile))
        using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
        {
            var records = csv.GetRecords<JeopardyQuestion>();
            foreach (var record in records)
            {
                var properties = new { question = record.Question, answer = record.Answer };
                batch.Add(properties);

                if (batch.Count == batchSize)
                {
                    await collection.Data.InsertMany(batch);
                    Console.WriteLine($"Imported {batch.Count} articles...");
                    batch.Clear();
                }
            }
        }

        if (batch.Any())
        {
            await collection.Data.InsertMany(batch);
            Console.WriteLine($"Imported remaining {batch.Count} articles...");
        }

        Console.WriteLine("Finished importing articles.");
        // END CSV streaming

        var result = await collection.Aggregate.OverAll(totalCount: true);
        Assert.Equal(1000, result.TotalCount);
    }

    // Helper class for CSV parsing
    private class JeopardyQuestion
    {
        public string Question { get; set; }
        public string Answer { get; set; }
    }
}

File: _includes/code/java-v6/src/test/java/ManageObjectsImportTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.Property;
import io.weaviate.client6.v1.api.collections.ReferenceProperty;
import io.weaviate.client6.v1.api.collections.VectorConfig;
import io.weaviate.client6.v1.api.collections.data.BatchReference;
import io.weaviate.client6.v1.api.collections.Vectors;
import io.weaviate.client6.v1.api.collections.WeaviateObject;
import io.weaviate.client6.v1.api.collections.data.InsertManyResponse;
import io.weaviate.client6.v1.api.collections.query.QueryReference;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

class ManageObjectsImportTest {

  private static WeaviateClient client;

  // A helper method to generate a deterministic UUID from a seed
  private static UUID generateUuid5(String seed) {
    return UUID.nameUUIDFromBytes(seed.getBytes(StandardCharsets.UTF_8));
  }

  @BeforeAll
  public static void beforeAll() throws IOException {
    // START INSTANTIATION-COMMON
    String openaiApiKey = System.getenv("OPENAI_API_KEY");
    assertThat(openaiApiKey).isNotBlank()
        .withFailMessage("Please set the OPENAI_API_KEY environment variable.");

    client = WeaviateClient.connectToLocal(
        config -> config.setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
    // END INSTANTIATION-COMMON
    // Download data file for streaming tests
    try (InputStream in = URI.create(
        "https://raw.githubusercontent.com/weaviate-tutorials/edu-datasets/main/jeopardy_1k.json")
        .toURL()
        .openStream()) {
      Files.copy(in, Paths.get("jeopardy_1k.json"));
    }
  }

  @AfterAll
  public static void afterAll() throws IOException {
    client.collections.deleteAll();
    Files.deleteIfExists(Paths.get("jeopardy_1k.json"));
  }

  @BeforeEach
  public void beforeEach() throws IOException {
    client.collections.deleteAll();
  }

  @Test
  void testBasicBatchImport() throws IOException {
    // Define and create the class
    client.collections.create("MyCollection",
        col -> col.vectorConfig(VectorConfig.selfProvided()));

    // START BasicBatchImportExample
    List<Map<String, Object>> dataRows = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      dataRows.add(Map.of("title", "Object " + (i + 1)));
    }

    var collection = client.collections.use("MyCollection");

    // The Java client uses insertMany for batching.
    // There is no direct equivalent of the Python client's stateful batch manager.
    // You collect objects and send them in a single request.
    // highlight-start
    var response = collection.data.insertMany(dataRows.toArray(new Map[0]));
    // highlight-end

    if (!response.errors().isEmpty()) {
      System.err
          .println("Number of failed imports: " + response.errors().size());
      System.err.println("First failed object: " + response.errors().get(0));
    }
    // END BasicBatchImportExample

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(5);

    client.collections.delete("MyCollection");
  }

  //@Test
  void testServerSideBatchImport() throws IOException {
    client.collections.create("MyCollection",
        col -> col.vectorConfig(VectorConfig.selfProvided()));

    // START ServerSideBatchImportExample
    List<Map<String, Object>> dataRows = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      dataRows.add(Map.of("title", "Object " + (i + 1)));
    }

    var collection = client.collections.use("MyCollection");

    // highlight-start
    // Use `batch.start()` for server-side batching. The client sends data
    // in batches at a rate controlled by the server.
    try (var batch = collection.batch.start()) {
      for (var dataRow : dataRows) {
        batch.add(WeaviateObject.of(
            obj -> obj.properties(dataRow)));
      }
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
    }
    // highlight-end
    // END ServerSideBatchImportExample

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(5);

    client.collections.delete("MyCollection");
  }

  @Test
  void testBatchImportWithID() throws IOException {
    client.collections.create("MyCollection",
        col -> col.vectorConfig(VectorConfig.selfProvided()));

    // START BatchImportWithIDExample
    List<WeaviateObject<Map<String, Object>>> dataObjects = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      Map<String, Object> dataRow = Map.of("title", "Object " + (i + 1));
      UUID objUuid = generateUuid5(dataRow.toString());

      dataObjects.add(WeaviateObject.<Map<String, Object>>of(
          obj -> obj.properties(dataRow).uuid(objUuid.toString())));
    }

    var collection = client.collections.use("MyCollection");

    // highlight-start
    var response = collection.data.insertMany(dataObjects);
    // highlight-end

    if (!response.errors().isEmpty()) {
      System.err
          .println("Number of failed imports: " + response.errors().size());
      System.err.println("First failed object: " + response.errors().get(0));
    }
    // END BatchImportWithIDExample

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(5);
    String lastUuid = dataObjects.get(4).uuid();
    assertThat(collection.data.exists(lastUuid)).isTrue();

    client.collections.delete("MyCollection");
  }

  @Test
  void testBatchImportWithVector() throws IOException {
    client.collections.create("MyCollection",
        col -> col.vectorConfig(VectorConfig.selfProvided()));

    // START BatchImportWithVectorExample
    List<WeaviateObject<Map<String, Object>>> dataObjects = new ArrayList<>();
    float[] vector = new float[10]; // Using a small vector for demonstration
    Arrays.fill(vector, 0.1f);

    for (int i = 0; i < 5; i++) {
      Map<String, Object> dataRow = Map.of("title", "Object " + (i + 1));
      UUID objUuid = generateUuid5(dataRow.toString());

      dataObjects.add(
          WeaviateObject.<Map<String, Object>>of(obj -> obj.properties(dataRow)
              .uuid(objUuid.toString())
              .vectors(Vectors.of(vector))));
    }

    var collection = client.collections.use("MyCollection");

    // highlight-start
    var response = collection.data.insertMany(dataObjects);
    // highlight-end

    if (!response.errors().isEmpty()) {
      System.err
          .println("Number of failed imports: " + response.errors().size());
      System.err.println("First failed object: " + response.errors().get(0));
    }
    // END BatchImportWithVectorExample

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(5);

    client.collections.delete("MyCollection");
  }

  @Test
  void testBatchImportWithCrossReference() throws IOException {
    client.collections.create("Publication",
        col -> col.properties(Property.text("title")));

    client.collections.create("Author",
        col -> col.properties(Property.text("name"))
            .references(ReferenceProperty.to("writesFor", "Publication")));

    var authors = client.collections.use("Author");
    var publications = client.collections.use("Publication");

    var from = authors.data.insert(Map.of("name", "Jane Austen"));
    var fromUuid = from.uuid();
    var targetUuid =
        publications.data.insert(Map.of("title", "Ye Olde Times")).uuid();

    // START BatchImportWithRefExample
    var collection = client.collections.use("Author");

    var response = collection.data
        .referenceAddMany(BatchReference.uuids(from, "writesFor", targetUuid));

    if (!response.errors().isEmpty()) {
      System.err
          .println("Number of failed imports: " + response.errors().size());
      System.err.println("First failed object: " + response.errors().get(0));
    }
    // END BatchImportWithRefExample

    var result = collection.query.fetchObjectById(fromUuid,
        q -> q.returnReferences(QueryReference.single("writesFor")));

    assertThat(result).isPresent();
    assertThat(result.get().references().get("writesFor")).isNotNull();
  }

  @Test
  void testImportWithNamedVectors() throws IOException {
    // Define and create the class
    client.collections.create("MyCollection",
        col -> col
            .vectorConfig(VectorConfig.selfProvided("title"),
                VectorConfig.selfProvided("body"))
            .properties(Property.text("title"), Property.text("body")));

    // START BatchImportWithNamedVectors
    // Prepare the data and vectors
    List<Map<String, Object>> dataRows = new ArrayList<>();
    List<float[]> titleVectors = new ArrayList<>();
    List<float[]> bodyVectors = new ArrayList<>();

    for (int i = 0; i < 5; i++) {
      dataRows
          .add(Map.of("title", "Object " + (i + 1), "body", "Body " + (i + 1)));

      float[] titleVector = new float[1536];
      Arrays.fill(titleVector, 0.12f);
      titleVectors.add(titleVector);

      float[] bodyVector = new float[1536];
      Arrays.fill(bodyVector, 0.34f);
      bodyVectors.add(bodyVector);
    }

    CollectionHandle<Map<String, Object>> collection =
        client.collections.use("MyCollection");

    List<WeaviateObject<Map<String, Object>>> objectsToInsert =
        new ArrayList<>();
    for (int i = 0; i < dataRows.size(); i++) {
      int index = i;
      objectsToInsert.add(
          // highlight-start
          // Use the Builder with the EXACT matching generic types
          WeaviateObject
              .<Map<String, Object>>of(v -> v.properties(dataRows.get(index))
                  .vectors(Vectors.of("title", titleVectors.get(index)))
                  .vectors(Vectors.of("body", bodyVectors.get(index)))));
      // highlight-end

    }

    // Insert the data using insertMany with the List
    // highlight-start
    InsertManyResponse response = collection.data.insertMany(objectsToInsert);
    // highlight-end

    // Check for errors
    if (!response.errors().isEmpty()) {
      System.err.printf("Number of failed imports: %d\n",
          response.errors().size());
      System.err.printf("First failed object error: %s\n",
          response.errors().get(0));
    }
    // END BatchImportWithNamedVectors
  }

  @Test
  void testJsonStreaming() throws IOException {
    client.collections.create("JeopardyQuestion");

    // START JSON streaming
    int batchSize = 100;
    List<Map<String, Object>> batch = new ArrayList<>(batchSize);
    var collection = client.collections.use("JeopardyQuestion");
    Gson gson = new Gson();

    System.out.println(
        "JSON streaming, to avoid running out of memory on large files...");
    try (JsonReader reader =
        new JsonReader(new FileReader("jeopardy_1k.json"))) {
      reader.beginArray();
      while (reader.hasNext()) {
        Map<String, String> obj = gson.fromJson(reader, Map.class);
        Map<String, Object> properties = new HashMap<>();
        properties.put("question", obj.get("Question"));
        properties.put("answer", obj.get("Answer"));
        batch.add(properties);

        if (batch.size() == batchSize) {
          collection.data.insertMany(batch.toArray(new Map[0]));
          System.out.println("Imported " + batch.size() + " articles...");
          batch.clear();
        }
      }
      reader.endArray();
    }

    if (!batch.isEmpty()) {
      collection.data.insertMany(batch.toArray(new Map[0]));
      System.out.println("Imported remaining " + batch.size() + " articles...");
    }

    System.out.println("Finished importing articles.");
    // END JSON streaming

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(1000);

    client.collections.delete("JeopardyQuestion");
  }

  @Test
  void testCsvStreaming() throws IOException {
    // Create a CSV file from the JSON for the test
    try (JsonReader reader = new JsonReader(new FileReader("jeopardy_1k.json"));
        java.io.FileWriter writer = new java.io.FileWriter("jeopardy_1k.csv")) {
      Gson gson = new Gson();
      reader.beginArray();
      writer.write("Question,Answer\n");
      while (reader.hasNext()) {
        Map<String, String> obj = gson.fromJson(reader, Map.class);
        writer.write(
            "\"" + obj.get("Question") + "\",\"" + obj.get("Answer") + "\"\n");
      }
      reader.endArray();
    }

    client.collections.create("JeopardyQuestion");

    // START CSV streaming
    int batchSize = 100;
    List<Map<String, Object>> batch = new ArrayList<>(batchSize);
    var collection = client.collections.use("JeopardyQuestion");

    System.out
        .println("CSV streaming to not load all records in RAM at once...");
    try (BufferedReader csvReader =
        new BufferedReader(new FileReader("jeopardy_1k.csv"))) {
      String line = csvReader.readLine(); // skip header
      while ((line = csvReader.readLine()) != null) {
        String[] data = line.split("\",\"");
        Map<String, Object> properties = new HashMap<>();
        properties.put("question", data[0].substring(1));
        properties.put("answer", data[1].substring(0, data[1].length() - 1));
        batch.add(properties);

        if (batch.size() == batchSize) {
          collection.data.insertMany(batch.toArray(new Map[0]));
          System.out.println("Imported " + batch.size() + " articles...");
          batch.clear();
        }
      }
    }

    if (!batch.isEmpty()) {
      collection.data.insertMany(batch.toArray(new Map[0]));
      System.out.println("Imported remaining " + batch.size() + " articles...");
    }

    System.out.println("Finished importing articles.");
    // END CSV streaming

    var result =
        collection.aggregate.overAll(agg -> agg.includeTotalCount(true));
    assertThat(result.totalCount()).isEqualTo(1000);

    Files.deleteIfExists(Paths.get("jeopardy_1k.csv"));
  }
}
